### DISCUSSION TITLE
`<RayGrab />` with blocked axis?

### DISCUSSION THREAD

**Discussion by JenniferFalkenstein** [pmndrs/xr#230] [Q&A] [answered]
Is it possible to use <RayGrab /> with a blocked axis, so that for instance an object can only be moved and dragged in x- and y-directions but not in z?
Or is there maybe a code example on how to code this feature myself? 😄

**Comment by alexbardyshev** [comment]
Hi @JenniferFalkenstein, were you able to find a solution to your question about restricting object movement to the x- and y-directions only?

**Reply by JenniferFalkenstein** [reply]
@alexbardyshev unfortunately I still don't have.
In my opinion this could be a nice feature for the react-xr library though, since there are several use-cases, where you may want to grab an item but move it around only in certain directions... 😕

**Reply by CodyJasonBennett** [reply, marked as answer]
If you create an issue, I'll be sure to get back to it.

Three's matrix system is a bit of a mess, but this importantly only works in local space, applying world-space transforms from the controller:

https://github.com/pmndrs/react-xr/blob/93d0e4c44676fd52c850dd80eafa8bc40bb83fef/src/Interactions.tsx#L222-L232

So we can memoize whichever axis shouldn't be moved and reapply it:

```diff
useFrame(() => {
  const controller = grabbingController.current
  const group = groupRef.current
  if (!controller) return
  
  // Memoize original values
+  const [x, y, z] = group.position

  group.applyMatrix4(previousTransform)
  group.applyMatrix4(controller.matrixWorld)

  // Keep original z value
+  group.position.z = z

  group.updateMatrixWorld()

  previousTransform.copy(controller.matrixWorld).invert()
})
```
